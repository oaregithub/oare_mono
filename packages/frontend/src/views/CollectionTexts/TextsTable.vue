<template>
  <v-data-table
    :headers="headers"
    :items="loading ? [] : texts"
    :loading="loading"
    :options.sync="searchOptions"
    :server-items-length="totalTexts"
    :footer-props="{
      'items-per-page-options': [10, 25, 50, 100],
    }"
  >
    <template v-slot:[`item.name`]="{ item }">
      <router-link v-if="item.hasEpigraphy" :to="`/epigraphies/${item.uuid}`">{{
        item.name
      }}</router-link>
      <span v-else>{{ item.name }}</span>
    </template>
  </v-data-table>
</template>

<script lang="ts">
import {
  defineComponent,
  ref,
  computed,
  watch,
  PropType,
} from '@vue/composition-api';
import { CollectionResponse, CollectionText } from '@/types/collections';
export default defineComponent({
  props: {
    loading: {
      type: Boolean,
      default: false,
    },
    texts: {
      type: Array as PropType<CollectionText[]>,
      required: true,
    },
    totalTexts: {
      type: Number,
      required: true,
    },
    page: {
      type: Number,
      default: 1,
    },
    rows: {
      type: Number,
      default: 10,
    },
  },
  setup(props, { emit }) {
    const headers = ref([
      {
        text: 'Text Name',
        value: 'name',
      },
    ]);

    const searchOptions = ref({
      page: props.page,
      itemsPerPage: props.rows,
    });

    watch(
      () => props.page,
      () => (searchOptions.value.page = props.page)
    );

    watch(
      () => searchOptions.value.page,
      () => {
        emit('update:page', searchOptions.value.page);
      }
    );

    watch(
      () => searchOptions.value.itemsPerPage,
      () => {
        emit('update:rows', searchOptions.value.itemsPerPage);
      }
    );

    return {
      headers,
      searchOptions,
    };
  },
});
</script>

<style></style>
