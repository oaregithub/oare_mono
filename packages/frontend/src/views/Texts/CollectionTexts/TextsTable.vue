<template>
  <v-data-table
    :headers="headers"
    :items="loading ? [] : texts"
    :loading="loading"
    :options.sync="searchOptions"
    :server-items-length="totalTexts"
    :footer-props="{
      'items-per-page-options': [10, 25, 50, 100],
    }"
  >
    <template v-slot:[`item.name`]="{ item }">
      <v-icon
        v-if="editText !== item.uuid"
        @click="toggleTextInfo(item.uuid, item)"
        class="test-pencil mr-4"
        >mdi-pencil</v-icon
      >
      <router-link v-if="item.hasEpigraphy" :to="`/epigraphies/${item.uuid}`">
        <span v-if="editText !== item.uuid">{{ item.name }}</span>
      </router-link>
      <span v-if="editText === item.uuid">
        <v-btn
          class="mr-3"
          color="primary"
          width="45px"
          @click="updateTextInfo(item)"
          >Save</v-btn
        >
        <v-btn color="info" width="45px" @click="cancelEditTextInfo(item)"
          >Cancel</v-btn
        >
      </span>
    </template>

    <template v-slot:[`item.excavation`]="{ item }">
      <v-row>
        <v-col cols="8" sm="6">
          <span v-if="editText !== item.uuid">
            {{ item.excavationPrefix }} {{ item.excavationNumber }}
          </span>
        </v-col>
      </v-row>
      <v-row v-if="editText === item.uuid">
        <v-col cols="8" sm="6">
          <v-text-field
            outlined
            v-model="item.excavationPrefix"
            label="Prefix"
            clearable
          ></v-text-field>
        </v-col>
        <v-col cols="8" sm="6">
          <v-text-field
            outlined
            v-model="item.excavationNumber"
            label="Prefix"
            clearable
          ></v-text-field>
        </v-col>
      </v-row>
    </template>

    <template v-slot:[`item.museum`]="{ item }">
      <v-row>
        <v-col cols="8" sm="6">
          <span v-if="editText !== item.uuid">
            {{ item.museumPrefix }} {{ item.museumNumber }}
          </span>
        </v-col>
      </v-row>
      <v-row v-if="editText === item.uuid">
        <v-col cols="8" sm="6">
          <v-text-field
            outlined
            v-model="item.museumPrefix"
            label="Prefix"
            clearable
          ></v-text-field>
        </v-col>
        <v-col cols="8" sm="6">
          <v-text-field
            outlined
            v-model="item.museumNumber"
            label="Prefix"
            clearable
          ></v-text-field>
        </v-col>
      </v-row>
    </template>

    <template v-slot:[`item.publication`]="{ item }">
      <v-row>
        <v-col cols="8" sm="6">
          <span v-if="editText !== item.uuid">
            {{ item.publicationPrefix }} {{ item.publicationNumber }}
          </span>
        </v-col>
      </v-row>
      <v-row v-if="editText === item.uuid">
        <v-col cols="4" sm="6">
          <v-text-field
            outlined
            v-model="item.publicationPrefix"
            label="Prefix"
            clearable
          ></v-text-field>
        </v-col>
        <v-col cols="4" sm="6">
          <v-text-field
            outlined
            v-model="item.publicationNumber"
            label="Prefix"
            clearable
          ></v-text-field>
        </v-col>
      </v-row>
    </template>
  </v-data-table>
</template>

<script lang="ts">
import { defineComponent, ref, watch, PropType } from '@vue/composition-api';
import { CollectionText } from '@oare/types';
import sl from '@/serviceLocator';

export interface OriginalTextInfo {
  excavationPrefix: string | null;
  excavationNumber: string | null;
  museumPrefix: string | null;
  museumNumber: string | null;
  primaryPublicationPrefix: string | null;
  primaryPublicationNumber: string | null;
}

export default defineComponent({
  props: {
    loading: {
      type: Boolean,
      default: false,
    },
    texts: {
      type: Array as PropType<CollectionText[]>,
      required: true,
    },
    totalTexts: {
      type: Number,
      required: true,
    },
    page: {
      type: Number,
      default: 1,
    },
    rows: {
      type: Number,
      default: 10,
    },
  },

  setup(props, { emit }) {
    const server = sl.get('serverProxy');
    const actions = sl.get('globalActions');

    const originalTextInfoObject = ref<OriginalTextInfo>({
      excavationPrefix: null,
      excavationNumber: null,
      museumPrefix: null,
      museumNumber: null,
      primaryPublicationPrefix: null,
      primaryPublicationNumber: null,
    });

    const headers = ref([
      {
        text: 'Text Name',
        value: 'name',
      },
      {
        text: 'Excavation Info',
        value: 'excavation',
      },
      {
        text: 'Museum Info',
        value: 'museum',
      },
      {
        text: 'Primary Publication Info',
        value: 'publication',
      },
    ]);

    const searchOptions = ref({
      page: props.page,
      itemsPerPage: props.rows,
    });

    const editText = ref<string>();

    const toggleTextInfo = function (uuid: string, item: any) {
      originalTextInfoObject.value.excavationPrefix = item.excavationPrefix;
      originalTextInfoObject.value.excavationNumber = item.excavationNumber;
      originalTextInfoObject.value.museumPrefix = item.museumPrefix;
      originalTextInfoObject.value.museumNumber = item.museumNumber;
      originalTextInfoObject.value.primaryPublicationPrefix =
        item.publicationPrefix;
      originalTextInfoObject.value.primaryPublicationNumber =
        item.publicationNumber;

      editText.value = uuid;
    };

    const cancelEditTextInfo = function (item: any) {
      item.excavationPrefix = originalTextInfoObject.value.excavationPrefix;
      item.excavationNumber = originalTextInfoObject.value.excavationNumber;
      item.museumPrefix = originalTextInfoObject.value.museumPrefix;
      item.museumNumber = originalTextInfoObject.value.museumNumber;
      item.publicationPrefix =
        originalTextInfoObject.value.primaryPublicationPrefix;
      item.publicationNumber =
        originalTextInfoObject.value.primaryPublicationNumber;

      editText.value = undefined;
    };

    const updateTextInfo = async (item: any) => {
      try {
        await server.updateTextInfo(
          item.uuid,
          item.excavationPrefix,
          item.excavationNumber,
          item.museumPrefix,
          item.museumNumber,
          item.publicationPrefix,
          item.publicationNumber
        );
        editText.value = undefined;
      } catch (err) {
        actions.showErrorSnackbar(
          'Error happened while editing text. Please try again.',
          err as Error
        );
      }
    };

    watch(
      () => props.page,
      () => (searchOptions.value.page = props.page)
    );

    watch(
      () => searchOptions.value.page,
      () => {
        emit('update:page', searchOptions.value.page);
      }
    );

    watch(
      () => searchOptions.value.itemsPerPage,
      () => {
        emit('update:rows', searchOptions.value.itemsPerPage);
      }
    );

    return {
      headers,
      searchOptions,
      editText,
      toggleTextInfo,
      cancelEditTextInfo,
      updateTextInfo,
    };
  },
});
</script>

<style></style>
