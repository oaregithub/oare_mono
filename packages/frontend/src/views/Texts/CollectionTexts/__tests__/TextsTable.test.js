import Vuetify from 'vuetify';
import VueCompositionApi from '@vue/composition-api';
import { createLocalVue, mount } from '@vue/test-utils';
import flushPromises from 'flush-promises';
import sl from '@/serviceLocator';
import TextsTable from '../TextsTable.vue';

const vuetify = new Vuetify();
const localVue = createLocalVue();
localVue.use(VueCompositionApi);

describe('TextsTable test', () => {
  const mockServer = {
    updateTextInfo: jest.fn().mockResolvedValue(),
  };

  const mockActions = {
    showErrorSnackbar: jest.fn(),
  };

  const mockStore = {
    getters: {
      permissions: [{ name: 'EDIT_TEXT_INFO' }],
      isAdmin: true,
    },
  };

  const setup = () => {
    sl.set('serverProxy', mockServer);
    sl.set('globalActions', mockActions);
    sl.set('store', mockStore);
  };

  beforeEach(setup);

  const texts = [
    {
      id: 1,
      uuid: '1',
      type: 'text',
      hasEpigraphy: true,
      name: 'CCT 3 31',
      excavationPrefix: null,
      excavationNumber: null,
      museumPrefix: null,
      museumNumber: null,
      publicationPrefix: null,
      publicationNumber: null,
    },
    {
      id: 2,
      uuid: '2',
      type: 'text',
      hasEpigraphy: true,
      name: 'CCT 1 12b',
      excavationPrefix: null,
      excavationNumber: null,
      museumPrefix: null,
      museumNumber: null,
      publicationPrefix: null,
      publicationNumber: null,
    },
  ];
  const mockProps = {
    loading: false,
    texts,
    totalTexts: 2,
  };

  const createWrapper = () =>
    mount(TextsTable, {
      localVue,
      vuetify,
      propsData: mockProps,
      stubs: ['router-link'],
    });

  it('matches snapshot', () => {
    expect(createWrapper()).toMatchSnapshot();
  });

  it('displays names of all texts', () => {
    const { getByText } = createWrapper();

    texts
      .map(t => t.name)
      .forEach(name => {
        expect(name);
      });
  });

  it('shows edit pencil button if user has EDIT_TEXT_INFO permission', async () => {
    const wrapper = createWrapper();
    await flushPromises();

    const addEditButton = wrapper.find('.test-pencil');
    expect(addEditButton.exists()).toBe(true);
  });

  it('does not show edit pencil button if user does not have EDIT_TEXT_INFO permission', async () => {
    sl.set('store', {
      getters: {
        ...mockStore.getters,
        permissions: [],
      },
    });
    const wrapper = createWrapper();
    await flushPromises();

    const addEditButton = wrapper.find('.test-pencil');
    expect(addEditButton.exists()).toBe(false);
  });

  it('successfully updates text info', async () => {
    const wrapper = createWrapper();
    await flushPromises();

    const editButton = wrapper.get('.test-pencil');
    await editButton.trigger('click');

    const excavationPrefixInput = wrapper.get('.excavationPrefix input');
    await excavationPrefixInput.setValue('excavationPrefixExample');

    const excavationNumberInput = wrapper.get('.excavationNumber input');
    await excavationNumberInput.setValue('excavationNumberExample');

    const museumPrefixInput = wrapper.get('.museumPrefix input');
    await museumPrefixInput.setValue('museumPrefixExample');

    const museumNumberInput = wrapper.get('.museumNumber input');
    await museumNumberInput.setValue('museumNumberExample');

    const publicationPrefixInput = wrapper.get('.publicationPrefix input');
    await publicationPrefixInput.setValue('publicationPrefixExample');

    const publicationNumberInput = wrapper.get('.publicationNumber input');
    await publicationNumberInput.setValue('publicationNumberExample');

    const saveButton = wrapper.get('.edit-text-save-btn');
    await saveButton.trigger('click');

    await flushPromises();

    expect(mockServer.updateTextInfo).toHaveBeenCalledWith(
      '1',
      'excavationPrefixExample',
      'excavationNumberExample',
      'museumPrefixExample',
      'museumNumberExample',
      'publicationPrefixExample',
      'publicationNumberExample'
    );
  });

  it('shows error snackbar on failed text info edit', async () => {
    sl.set('serverProxy', {
      ...mockServer,
      updateTextInfo: jest.fn().mockRejectedValue(),
    });

    const wrapper = createWrapper();
    await flushPromises();

    const editButton = wrapper.get('.test-pencil');
    await editButton.trigger('click');

    const excavationPrefixInput = wrapper.get('.excavationPrefix input');
    await excavationPrefixInput.setValue('excavationPrefixExample');

    const excavationNumberInput = wrapper.get('.excavationNumber input');
    await excavationNumberInput.setValue('excavationNumberExample');

    const museumPrefixInput = wrapper.get('.museumPrefix input');
    await museumPrefixInput.setValue('museumPrefixExample');

    const museumNumberInput = wrapper.get('.museumNumber input');
    await museumNumberInput.setValue('museumNumberExample');

    const publicationPrefixInput = wrapper.get('.publicationPrefix input');
    await publicationPrefixInput.setValue('publicationPrefixExample');

    const publicationNumberInput = wrapper.get('.publicationNumber input');
    await publicationNumberInput.setValue('publicationNumberExample');

    const saveButton = wrapper.get('.edit-text-save-btn');
    await saveButton.trigger('click');

    await flushPromises();

    expect(mockActions.showErrorSnackbar).toHaveBeenCalled();
  });
});
