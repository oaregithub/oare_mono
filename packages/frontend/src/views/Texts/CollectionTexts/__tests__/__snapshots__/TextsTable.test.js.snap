// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextsTable test matches snapshot 1`] = `
<div class="v-data-table theme--light" totaltexts="2">
  <div class="v-data-table__wrapper">
    <table>
      <colgroup>
        <col class="">
        <col class="">
        <col class="">
        <col class="">
      </colgroup>
      <thead class="v-data-table-header">
        <tr>
          <th role="columnheader" scope="col" aria-label="Text Name: Not sorted. Activate to sort ascending." aria-sort="none" class="text-start sortable"><span>Text Name</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
          <th role="columnheader" scope="col" aria-label="Excavation Info: Not sorted. Activate to sort ascending." aria-sort="none" class="text-start sortable"><span>Excavation Info</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
          <th role="columnheader" scope="col" aria-label="Museum Info: Not sorted. Activate to sort ascending." aria-sort="none" class="text-start sortable"><span>Museum Info</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
          <th role="columnheader" scope="col" aria-label="Primary Publication Info: Not sorted. Activate to sort ascending." aria-sort="none" class="text-start sortable"><span>Primary Publication Info</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
        </tr>
      </thead>
      <tbody>
        <tr class="">
          <td class="text-start"><button type="button" class="v-icon notranslate test-pencil mr-4 v-icon--link mdi mdi-pencil theme--light"></button>
            <router-link-stub to="/epigraphies/1">
              CCT 3 31
            </router-link-stub>
            <!---->
          </td>
          <td class="text-start">
            <div class="row">
              <div class="col-sm-6 col-8"><span>   </span></div>
            </div>
            <!---->
          </td>
          <td class="text-start">
            <div class="row">
              <div class="col-sm-6 col-8"><span>   </span></div>
            </div>
            <!---->
          </td>
          <td class="text-start">
            <div class="row">
              <div class="col-sm-6 col-8"><span>

        </span></div>
            </div>
            <!---->
          </td>
        </tr>
        <tr class="">
          <td class="text-start"><button type="button" class="v-icon notranslate test-pencil mr-4 v-icon--link mdi mdi-pencil theme--light"></button>
            <router-link-stub to="/epigraphies/2">
              CCT 1 12b
            </router-link-stub>
            <!---->
          </td>
          <td class="text-start">
            <div class="row">
              <div class="col-sm-6 col-8"><span>   </span></div>
            </div>
            <!---->
          </td>
          <td class="text-start">
            <div class="row">
              <div class="col-sm-6 col-8"><span>   </span></div>
            </div>
            <!---->
          </td>
          <td class="text-start">
            <div class="row">
              <div class="col-sm-6 col-8"><span>

        </span></div>
            </div>
            <!---->
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
